consensus/src/dag/round_state.rs

```diff
@@ -0,0 +1,154 @@
+                .await
+                .expect("Not able to send new round event");
+        }
+    }
+
+    fn current_round(&self) -> Round {
+        self.current_round
+    }
+
+    fn set_current_round(&mut self, round: Round) {
+        assert!(self.current_round < round);
+        self.current_round = round;
+    }
+}
+
+/// More sophisticated strategy to move round forward given 2f+1 strong links
+/// Delay if backpressure is triggered. (TODO)
+/// Move as soon as 3f+1 is ready. (TODO: make it configurable)
+/// Move if minimal wait time is reached. (TODO: make it configurable)
+pub struct AdaptiveResponsiveRound {
+    current_round: Round,
+    epoch_state: Arc<EpochState>,
+    start_time: Instant,
+    minimal_wait_time: Duration,
+    event_sender: tokio::sync::mpsc::Sender<Round>,
+}
+
+impl AdaptiveResponsiveRound {
+    pub fn new(
+        event_sender: tokio::sync::mpsc::Sender<Round>,
+        epoch_state: Arc<EpochState>,
+        minimal_wait_time: Duration,
+    ) -> Self {
+        Self {
+            current_round: 0,
+            epoch_state,
+            start_time: Instant::now(),
+            minimal_wait_time,
+            event_sender,
+        }
+    }
+
+    async fn send_new_round(&self, round: Round) {
+        self.event_sender
+            .send(round)
+            .await
+            .expect("Unable to send new round event");
+    }
+}
+
+#[async_trait]
+impl RoundState for AdaptiveResponsiveRound {
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        strong_links: Vec<NodeCertificate>,
+    ) {
+        let new_round = highest_strong_links_round + 1;
+        match self.current_round.cmp(&highest_strong_links_round) {
+            // We're behind, move forward immediately
+            Ordering::Less => self.send_new_round(new_round).await,
+            // Decide when to move to next round
+            Ordering::Equal => {
+                let voting_power = self
+                    .epoch_state
+                    .verifier
+                    .sum_voting_power(strong_links.iter().map(|cert| cert.metadata().author()))
+                    .expect("Unable to sum voting power from strong links");
+                // voting power == 3f+1 or pass minimal wait time
+                let duration_since_start = Instant::now().duration_since(self.start_time);
+                if voting_power == self.epoch_state.verifier.total_voting_power()
+                    || duration_since_start >= self.minimal_wait_time
+                {
+                    self.send_new_round(new_round).await;
+                } else {
+                    // wait until minimal time reaches before sending
+                    let sender = self.event_sender.clone();
+                    let wait_time = self.minimal_wait_time - duration_since_start;
+                    tokio::spawn(async move {
+                        tokio::time::sleep(wait_time).await;
```

Comment:
What if we get more votes in the meantime?

Reply:
Votes are collected when processing the actual new round event in dag driver so it'll be included